File: Tests/Strata.Foundation.CoreTest/test/typescript/Utility/SuppliedTest.ts
import "jest"
import {Supplied,Optional,NullPointerException} from "strata.foundation.core/Utility";

describe(
    "Supplied",
    () =>
    {
        it(
            "testSuccessCreation",
            () =>
            {
                const result: Supplied<string> = Supplied.of("testMethod","testOutput");

                expect(result.isSuccess()).toBeTruthy();
                expect(result.isFailure()).toBeFalsy();
                expect(result.getMethod()).toBe("testMethod");
                expect(result.getOutput().isPresent()).toBeTruthy();
                expect(result.getOutput().get()).toBe("testOutput");
                expect(result.getException().isEmpty()).toBeTruthy();
            });
        it(
            "testFailureCreation",
            () =>
            {
                const error: Error = new Error("test error");
                const result: Supplied<string> = Supplied.of<string>("testMethod",error);

                expect(result.isSuccess()).toBeFalsy();
                expect(result.isFailure()).toBeTruthy();
                expect(result.getMethod()).toBe("testMethod");
                expect(result.getOutput().isEmpty()).toBeTruthy();
                expect(result.getException().isPresent()).toBeTruthy();
                expect(result.getException().get().message).toBe("test error");
            });
        it(
            "testNullMethod",
            () =>
            {
                try
                {
                    Supplied.of(null as unknown as string,"testOutput");
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NullPointerException).toBeTruthy();
                }
            });
        it(
            "testNullOutput",
            () =>
            {
                const result: Supplied<string> = Supplied.of("testMethod",null as unknown as string);

                expect(result.isSuccess()).toBeFalsy();
                expect(result.isFailure()).toBeTruthy();
                expect(result.getOutput().isEmpty()).toBeTruthy();
                expect(result.getException().isPresent()).toBeTruthy();
            });
    });

File: Tests/Strata.Foundation.CoreTest/test/java/strata/foundation/core/utility/SuppliedTest.java
//////////////////////////////////////////////////////////////////////////////
// SuppliedTest.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

@Tag("CommitStage")
public
class SuppliedTest
{
    private Supplied<String> success;
    private Supplied<String> failure;

    @BeforeEach
    public void
    setUp()
    {
        success = Supplied.of("testMethod","testOutput");
        failure = Supplied.of("testMethod",new RuntimeException("test error"));
    }

    @Test
    public void
    testSuccessCreation()
    {
        assertTrue(success.isSuccess());
        assertFalse(success.isFailure());
        assertEquals("testMethod",success.getMethod());
        assertTrue(success.getOutput().isPresent());
        assertEquals("testOutput",success.getOutput().get());
        assertFalse(success.getThrowable().isPresent());
    }

    @Test
    public void
    testFailureCreation()
    {
        assertFalse(failure.isSuccess());
        assertTrue(failure.isFailure());
        assertEquals("testMethod",failure.getMethod());
        assertFalse(failure.getOutput().isPresent());
        assertTrue(failure.getThrowable().isPresent());
        assertEquals("test error",failure.getThrowable().get().getMessage());
    }

    @Test
    public void
    testNullMethod()
    {
        try
        {
            Supplied.of(null,"testOutput");
            fail("Should have thrown NullPointerException");
        }
        catch (NullPointerException e) {}
    }

    @Test
    public void
    testNullOutput()
    {
        Supplied<String> result = Supplied.of("testMethod",(String)null);

        assertFalse(result.isSuccess());
        assertTrue(result.isFailure());
        assertFalse(result.getOutput().isPresent());
        assertTrue(result.getThrowable().isPresent());
    }

    @Test
    public void
    testNullThrowable()
    {
        Supplied<String> result = Supplied.of("testMethod",(Throwable)null);

        assertFalse(result.isSuccess());
        assertTrue(result.isFailure());
        assertTrue(result.getThrowable().isPresent());
    }
}

//////////////////////////////////////////////////////////////////////////////
